client/exec.go

```diff
@@ -0,0 +1,305 @@
+	Terminal bool
+
+	// True to separate the process's stderr into a separate websocket. The
+	// default is to send combined stdout+stderr on a single websocket.
+	//
+	// Note: currently only the combinations Terminal=true, Stderr=false and
+	// Terminal=false, Stderr=true are supported.
+	Stderr bool
+
+	// Initial terminal width and height (only apply if Terminal is true).
+	// If not specified, the Pebble server uses the target's default (usually
+	// 80x25). When using the "pebble exec" CLI, these are set to the host's
+	// terminal size automatically.
+	Width  int
+	Height int
+}
+
+// ExecAdditionalArgs are additional control arguments for the Exec call.
+type ExecAdditionalArgs struct {
+	// Standard input, output, and error.
+	Stdin  io.ReadCloser
+	Stdout io.Writer
+	Stderr io.Writer
+
+	// Control message handler (for window resizing and signal forwarding)
+	Control func(conn WebsocketWriter)
+
+	// Channel that will be closed when all data operations are done
+	DataDone chan bool
+}
+
+// Exec starts a command execution with the given options and additional
+// control arguments, returning the execution's change ID.
+func (client *Client) Exec(opts *ExecOptions, args *ExecAdditionalArgs) (string, error) {
+	if opts.Terminal && opts.Stderr {
+		return "", errors.New("separate stderr not currently supported in terminal mode")
+	}
+	if !opts.Terminal && !opts.Stderr {
+		return "", errors.New("combined stderr not currently supported in non-terminal mode")
+	}
+
+	var timeoutStr string
+	if opts.Timeout != 0 {
+		timeoutStr = opts.Timeout.String()
+	}
+
+	var payload = struct {
+		Command     []string          `json:"command"`
+		Environment map[string]string `json:"environment"`
+		WorkingDir  string            `json:"working-dir"`
+		Timeout     string            `json:"timeout"`
+		UserID      *int              `json:"user-id"`
+		User        string            `json:"user"`
+		GroupID     *int              `json:"group-id"`
+		Group       string            `json:"group"`
+		Terminal    bool              `json:"terminal"`
+		Stderr      bool              `json:"stderr"`
+		Width       int               `json:"width"`
+		Height      int               `json:"height"`
+	}{
+		Command:     opts.Command,
+		Environment: opts.Environment,
+		WorkingDir:  opts.WorkingDir,
+		Timeout:     timeoutStr,
+		UserID:      opts.UserID,
+		User:        opts.User,
+		GroupID:     opts.GroupID,
+		Group:       opts.Group,
+		Terminal:    opts.Terminal,
+		Stderr:      opts.Stderr,
+		Width:       opts.Width,
+		Height:      opts.Height,
+	}
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return "", err
+	}
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	resultBytes, changeID, err := client.doAsyncFull("POST", "/v1/exec", nil, headers, &body)
+	if err != nil {
+		return "", err
+	}
+	var result struct {
+		WebsocketIDs map[string]string `json:"websocket-ids"`
```

Comment:
I've done a bit of digging, and I think this should be `websocket-ids` as I have it. Most things with just the noun, like `tasks` or 'paths' are an actual list of objects, so the `-ids` suffix makes explicit it's just a bunch of IDs. This is somewhat true in snapd too, for example `snap-id` and `account-id` and `common-ids` are string IDs, whereas `changes` and `tasks` and `plugs` and `slots` are objects (however, it's not consistent, as it also has `snaps` and `users` as `[]string` -- but I believe that's for an input).

I think there's a difference here between input parameters and results. In the "start service" endpoint, the input field `services` is a `[]string` of service names, but it's called just `services` rather than `service-names`, because inputs are so often IDs/names. Whereas in a response `services` would be service objects.

Reply:
Alright, sounds good then.